pkg/admission/customresourcevalidation/authentication/validate_authentication.go

```diff
@@ -86,25 +86,58 @@ func (authenticationV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOld
 		return errs
 	}
 
-	// TODO validate the obj.  remember that status validation should *never* fail on spec validation errors.
 	errs = append(errs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
 	errs = append(errs, validateAuthenticationStatus(obj.Status)...)
 
 	return errs
 }
 
+func validateAuthenticationSpecCreate(spec configv1.AuthenticationSpec) field.ErrorList {
+	return validateAuthenticationSpec(spec)
+}
+
+func validateAuthenticationSpecUpdate(newspec, oldspec configv1.AuthenticationSpec) field.ErrorList {
+	return validateAuthenticationSpec(newspec)
+}
+
 func validateAuthenticationSpec(spec configv1.AuthenticationSpec) field.ErrorList {
 	errs := field.ErrorList{}
+	specField := field.NewPath("spec")
+
+	switch spec.Type {
+	case configv1.AuthenticationTypeNone:
+	case configv1.AuthenticationTypeIntegratedOAuth:
+	default:
+		errs = append(errs, field.NotSupported(specField.Child("type"),
+			spec.Type,
+			[]string{string(configv1.AuthenticationTypeNone), string(configv1.AuthenticationTypeIntegratedOAuth)},
+		))
+	}
 
-	// TODO
+	if oauthMetaErr := crvalidation.ValidateConfigMapReferenceName(specField.Child("oauthMetadata", "name"), spec.OAuthMetadata, false); oauthMetaErr != nil {
+		errs = append(errs, oauthMetaErr)
+	}
+
+	// validate the secret names in WebhookTokenAuthenticators
+	for i, wh := range spec.WebhookTokenAuthenticators {
+		if whErr := crvalidation.ValidateSecretReferenceName(
+			specField.Child("webhookTokenAuthenticators").Index(i).Child("kubeConfig", "name"),
+			wh.KubeConfig,
+			true,
+		); whErr != nil {
+			errs = append(errs, whErr)
+		}
+	}
 
 	return errs
 }
 
 func validateAuthenticationStatus(status configv1.AuthenticationStatus) field.ErrorList {
 	errs := field.ErrorList{}
 
-	// TODO
+	if oauthMetaErr := crvalidation.ValidateConfigMapReferenceName(field.NewPath("status", "integratedOAuthMetadata", "name"), status.IntegratedOAuthMetadata, false); oauthMetaErr != nil {
+		errs = append(errs, oauthMetaErr)
```

Comment:
Blindly append here by making `ValidateConfigMapReference` return a slice (applies to other places as well):

`errs = append(errs, crvalidation.ValidateConfigMapReference()...`